runtime/serde/serde-xml/common/src/aws/smithy/kotlin/runtime/serde/xml/XmlDeserializer.kt

```diff
@@ -328,6 +331,63 @@ internal class XmlStructDeserializer(
     }
 }
 
+/**
+ * Deserializes XML structures that are unwrapped
+ * See: https://smithy.io/2.0/aws/customizations/s3-customizations.html#aws-customizations-s3unwrappedxmloutput-trait
+ *
+ * @param objDescriptor associated [SdkObjectDescriptor] which represents the expected structure
+ * @param reader underlying [XmlStreamReader] from which tokens are read
+ */
+internal class UnwrappedXmlStructDeserializer(
+    private val objDescriptor: SdkObjectDescriptor,
+    private val reader: XmlStreamReader,
+) : Deserializer.FieldIterator {
+    private val FIELD_INDEX: Int = 0
+    override fun findNextFieldIndex(): Int? = if (reader.nextToken() is XmlToken.Text) FIELD_INDEX else null
+
+    private fun <T> deserializeValue(transform: ((String) -> T)): T {
+        val value = reader.takeCurrentAs<XmlToken.Text>().value ?: ""
+        return transform(value)
+    }
+
+    override fun skipValue() = reader.skipNext()
+
+    override fun deserializeByte(): Byte = deserializeValue { it.toIntOrNull()?.toByte() ?: throw DeserializationException("Unable to deserialize $it") }
+
+    override fun deserializeInt(): Int = deserializeValue { it.toIntOrNull() ?: throw DeserializationException("Unable to deserialize $it") }
+
+    override fun deserializeShort(): Short = deserializeValue { it.toIntOrNull()?.toShort() ?: throw DeserializationException("Unable to deserialize $it") }
+
+    override fun deserializeLong(): Long = deserializeValue { it.toLongOrNull() ?: throw DeserializationException("Unable to deserialize $it") }
+
+    override fun deserializeFloat(): Float = deserializeValue { it.toFloatOrNull() ?: throw DeserializationException("Unable to deserialize $it") }
+
+    override fun deserializeDouble(): Double = deserializeValue { it.toDoubleOrNull() ?: throw DeserializationException("Unable to deserialize $it") }
+
+    override fun deserializeString(): String = deserializeValue { it }
+
+    override fun deserializeBoolean(): Boolean = deserializeValue { it.toBoolean() }
```

Comment:
**Correctness**: I see a lot of duplicate code between `UnwrappedXmlStructDeserializer` and `XmlStructDeserializer`, which we should avoid. Let's either:
* make `XmlStructDeserializer` an `open class` so that the `UnwrappedXmlStructDeserializer` can inherit it and override the specific things that need to change –**or**–
* make an abstract base class for XML deserialization and make both `XmlStructDeserializer` and  `UnwrappedXmlStructDeserializer` inherit it, implementing only the functionality they individually need

Reply:
Do we even need to override this? Can't this just be handled in the initial `deserializeStruct` method which is doing the token advance to the next element? 🤔